File: src/main/java/me/cephetir/skyskipped/features/impl/discordrpc/DiscordRPCManager.java
/*
 *
 * DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
 * Version 2, December 2004
 *
 * Copyright (C) 2021 Cephetir
 *
 * Everyone is permitted to copy and distribute verbatim or modified
 * copies of this license document, and changing it is allowed as long
 * as the name is changed.
 *
 * DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
 * TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION
 *
 *  0. You just DO WHAT THE FUCK YOU WANT TO.
 */

package me.cephetir.skyskipped.features.impl.discordrpc;

import com.jagrosh.discordipc.IPCClient;
import com.jagrosh.discordipc.IPCListener;
import com.jagrosh.discordipc.entities.RichPresence;
import lombok.Getter;
import me.cephetir.skyskipped.SkySkipped;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;
import org.json.JSONObject;

import java.time.OffsetDateTime;

@Getter
public class DiscordRPCManager implements IPCListener {
    private static final long APPLICATION_ID = 867366183057752094L;
    private static final long UPDATE_PERIOD = 4200L;

    private IPCClient client;
    private String detailsLine;
    private String stateLine;
    private OffsetDateTime startTimestamp;
    private boolean connected;

    public void start() {
        try {
            if (isActive()) return;

            SkySkipped.getLogger().info("Starting Discord RP...");
            stateLine = "Starting...";
            detailsLine = "";
            startTimestamp = OffsetDateTime.now();
            client = new IPCClient(APPLICATION_ID);
            client.setListener(this);
            try {
                client.connect();
            } catch (Exception e) {
                SkySkipped.getLogger().error("Failed to connect to Discord RPC: " + e.getMessage());
            }
        } catch (Throwable ex) {
            SkySkipped.getLogger().error("DiscordRP has thrown an unexpected error while trying to start...");
            ex.printStackTrace();
        }
    }

    public void stop() {
        if (isActive()) {
            client.close();
            connected = false;
        }
    }

    public boolean isActive() {
        return client != null && connected;
    }

    public void updatePresence() {
        RichPresence presence = new RichPresence.Builder()
                .setState(stateLine)
                .setDetails(detailsLine)
                .setStartTimestamp(startTimestamp)
                .setLargeImage("large", "SkySkipped v" + SkySkipped.VERSION)
                .build();
        client.sendRichPresence(presence);
    }

    public void setStateLine(String status) {
        this.stateLine = status;
    }

    public void setDetailsLine(String status) {
        this.detailsLine = status;
    }

    @Override
    public void onReady(IPCClient client) {
        SkySkipped.getLogger().info("Discord RPC started");
        connected = true;
        MinecraftForge.EVENT_BUS.register(this);
    }

    private int tickCounter;

    @SubscribeEvent
    public void onClientTick(TickEvent.ClientTickEvent event) {
        tickCounter++;
        if (tickCounter % 200 == 0) {
            updatePresence();
            tickCounter = 0;
        }
    }

    @Override
    public void onClose(IPCClient client, JSONObject json) {
        SkySkipped.getLogger().info("Discord RPC closed");
        this.client = null;
        connected = false;
        MinecraftForge.EVENT_BUS.unregister(this);
    }

    @Override
    public void onDisconnect(IPCClient client, Throwable t) {
        SkySkipped.getLogger().info("Discord RPC disconnected");
        this.client = null;
        connected = false;
        MinecraftForge.EVENT_BUS.unregister(this);
    }
}

File: src/main/kotlin/me/cephetir/skyskipped/features/impl/discordrpc/DiscordRPCManager.kt
/*
 * SkySkipped - Hypixel Skyblock QOL mod
 * Copyright (C) 2023  Cephetir
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package me.cephetir.skyskipped.features.impl.discordrpc

import com.jagrosh.discordipc.entities.RichPresence
import com.jagrosh.discordipc.exceptions.NoDiscordClientException
import me.cephetir.bladecore.core.listeners.SkyblockIsland
import me.cephetir.bladecore.core.listeners.SkyblockListener
import me.cephetir.bladecore.discord.AbstractDiscordIPC
import me.cephetir.bladecore.utils.TextUtils.stripColor
import me.cephetir.bladecore.utils.threading.BackgroundScope
import me.cephetir.skyskipped.SkySkipped
import me.cephetir.skyskipped.config.Config
import me.cephetir.skyskipped.utils.mc
import java.time.OffsetDateTime

class DiscordRPCManager : AbstractDiscordIPC() {
    override val APP_ID = 1102633365858947162L
    override val rpcBuilder: RichPresence.Builder = RichPresence.Builder()
        .setLargeImage("large", "SkySkipped v${SkySkipped.VERSION}")

    override fun handleConnect() {
        SkySkipped.logger.info("Starting Discord RPC...")
        try {
            ipc!!.connect()
            rpcBuilder.setStartTimestamp(OffsetDateTime.now())
            val richPresence = rpcBuilder.build()
            ipc!!.sendRichPresence(richPresence)
            BackgroundScope.launchLooping(updateJob)
            SkySkipped.logger.info("Discord RPC initialised successfully!")
        } catch (e: NoDiscordClientException) {
            SkySkipped.logger.error("No discord client found for RPC, stopping")
        }
    }

    override fun handleDisconnect() {
        SkySkipped.logger.info("Shutting down Discord RPC...")
        BackgroundScope.cancel(updateJob)
        ipc!!.close()
    }

    override fun updateRPC(): RichPresence = rpcBuilder
        .setDetails(getLine(Config.drpcDetail.value))
        .setState(getLine(if (Config.drpcState.value == 4) 5 else Config.drpcState.value))
        .build()

    private fun getLine(line: Int): String = when (line) {
        0 ->
            if (SkyblockListener.island == SkyblockIsland.Unknown)
                if (mc.isIntegratedServerRunning) "Singleplayer" else mc.currentServerData?.serverIP ?: "Main Menu"
            else SkyblockListener.island.formattedName

        1 -> mc.session.username
        2 -> if (mc.isIntegratedServerRunning) "Singleplayer" else mc.currentServerData?.serverIP ?: "Main Menu"
        3 -> mc.thePlayer?.heldItem?.displayName?.stripColor()?.trim() ?: "Nothing"
        4 -> Config.drpcText.value
        5 -> Config.drpcText2.value
        else -> ""
    }
}